src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/stores/ResourceFormStore.js

```diff
@@ -87,15 +88,163 @@ export default class ResourceFormStore extends AbstractFormStore implements Form
+                    const oldChildData = oldData[key].length > childKey ? oldData[key][childKey] || {} : {};
+
+                    const oldChildSchema = oldTypes[oldChildData.type]?.form;
+                    const newChildSchema = newTypes[newChildData.type].form;
+
+                    if (Object.keys(oldChildData).length === 0 && !oldChildSchema) {
+                        // set originData
+                        if (!result[key]) {
+                            result[key] = [];
+                        }
+                        result[key][childKey] = newChildData;
+
+                        continue;
+                    }
+
+                    if (!result[key]){
+                        result[key] = [];
+                    }
+                    if (!(oldChildData.type in newTypes)) {
+                        result[key][childKey] = this.calculateDifference(
+                            oldChildSchema,
+                            newChildSchema,
+                            oldChildData,
+                            newChildData,
+                            parentPath.concat([key, childKey])
+                        );
+                    }
+                }
+
+                continue;
+            }
+
+            if (!oldData[key] || newSchema[key].type !== oldSchema[key]?.type) {
+                // set originData
+                result[key] = newData[key];
+            }
+        }
+
+        return result;
+    }
+
+    removeObsoleteData(oldSchema: Schema, newSchema: Schema, data: Object, parentPath: string[] = []) {
+        if (!oldSchema || !newSchema) {
+            return;
+        }
+
+        for (const key in oldSchema) {
+            const {items: newItems, type: newType, types: newTypes} = newSchema[key] || {};
+            const {items: oldItems, type: oldType, types: oldTypes} = oldSchema[key];
+
+            if (newType === SECTION_TYPE && newItems &&
+                oldType === SECTION_TYPE && oldItems) {
+                this.removeObsoleteData(
+                    oldItems,
+                    newItems,
+                    data,
+                    parentPath
+                );
+                continue;
+            }
+
+            if (newTypes && oldTypes
+                && Object.keys(newTypes).length > 0 && Object.keys(oldTypes).length > 0
+                && data[key] && isArrayLike(data[key])
+            ) {
+                for (const childKey of data[key].keys()) {
+                    const childData = data[key][childKey];
+                    const oldChildSchema = oldTypes[childData.type]?.form;
+                    let newChildSchema = newTypes[childData.type]?.form;
+
+                    if (!newChildSchema) {
+                        const defaultType = newSchema[key]?.defaultType;
+                        if (defaultType) {
+                            newChildSchema = newTypes[defaultType].form;
+                            //set default type
```

Comment:
would expect that this method only removes things. shouldnt this be set in the `calculateDifference` (`mergeData`) method? 🙈

Reply:
I'm also not happy with the method name, but if we split setting the `defaultValue` for unknown blocks and removing the obsolete properties into two methods, we would also have to iterate twice through the `oldSchema`.